tests/simpleKVBC/TesterCRE/main.cpp

```diff
@@ -129,60 +136,162 @@ ICommunication* createCommunication(const ClientConfig& cc,
+                            uint64_t init_update_block,
+                            uint64_t init_tls_update_block,
+                            std::shared_ptr<concord::secretsmanager::ISecretsManagerImpl> sm)
+      : clientId_{clientId}, key_path_{key_path}, tls_key_path_{certFolder}, sm_{sm} {
     init_last_update_block_ = init_update_block;
+    init_last_tls_update_block_ = init_tls_update_block;
   }
   bool validate(const State& state) const {
-    if (state.blockid < init_last_update_block_) return false;
     concord::messages::ClientStateReply crep;
     concord::messages::deserialize(state.data, crep);
-    return std::holds_alternative<concord::messages::ClientKeyExchangeCommand>(crep.response);
+    if (std::holds_alternative<concord::messages::ClientKeyExchangeCommand>(crep.response)) {
+      concord::messages::ClientKeyExchangeCommand command =
+          std::get<concord::messages::ClientKeyExchangeCommand>(crep.response);
+      if (command.tls && state.blockid >= init_last_tls_update_block_) return true;
+      if (!command.tls && state.blockid >= init_last_update_block_) return true;
+    }
+    return false;
   };
   bool execute(const State& state, WriteState& out) {
-    LOG_INFO(getLogger(), "execute key exchange request");
     concord::messages::ClientStateReply crep;
     concord::messages::deserialize(state.data, crep);
     concord::messages::ClientKeyExchangeCommand command =
         std::get<concord::messages::ClientKeyExchangeCommand>(crep.response);
+    if (!command.tls) return executeTransactionSigningKeyExchange(out);
+    if (command.tls) return executeTlsKeyExchange(out);
+    return false;
+  }
 
+ private:
+  logging::Logger getLogger() {
+    static logging::Logger logger_(logging::getLogger("concord.client.reconfiguration.testerCre.KeyExchangeHandler"));
+    return logger_;
+  }
+
+  bool executeTransactionSigningKeyExchange(WriteState& out) {
+    LOG_INFO(getLogger(), "execute transaction signing key exchange request");
     // Generate new key pair
     auto hex_keys = concord::util::crypto::Crypto::instance().generateRsaKeyPair(
         2048, concord::util::crypto::KeyFormat::HexaDecimalStrippedFormat);
     auto pem_keys = concord::util::crypto::Crypto::instance().RsaHexToPem(hex_keys);
 
     concord::messages::ReconfigurationRequest rreq;
     concord::messages::ClientExchangePublicKey creq;
+    fs::path enc_file_path = key_path_;
+    enc_file_path += ".enc";
+    bool enc = false;
+    std::fstream f(enc_file_path.string());
+    if (f.good()) {
+      enc_file_path += ".new";
+      enc = true;
+      sm_->encryptFile(enc_file_path.string(), pem_keys.first);
+    }
     fs::path new_key_path = key_path_;
-    new_key_path += ".new";
-    sm_->encryptFile(new_key_path.string(), pem_keys.first);
+    bool non_enc = false;
+    std::fstream f2(new_key_path);
+    if (f2.good()) {
+      new_key_path += ".new";
+      plain_sm_.encryptFile(new_key_path.string(), pem_keys.first);
```

Comment:
pem_keys.second?

Reply:
No, first is the private key and second is the public key. We only want to save (encrypted or not) the public key